controller/aks-cluster-config-handler.go

```diff
@@ -0,0 +1,684 @@
+		}
+	}
+	if config.Spec.KubernetesVersion != "" {
+		if !isKubernetesVersion(config.Spec.KubernetesVersion) {
+			return fmt.Errorf("Kubernetes version [%s] is not supported for cluster [%s]", config.Spec.KubernetesVersion, config.Spec.ClusterName)
+		}
+	}
+
+	return nil
+}
+
+// buildUpstreamClusterState
+func buildUpstreamClusterState(ctx context.Context, config *aksv1.AKSClusterConfig) (*aksv1.AKSClusterConfigSpec, error) {
+	upstreamSpec := &aksv1.AKSClusterConfigSpec{}
+
+	upstreamSpec.Imported = true
+
+	azureAuthorizer, _ := newClientAuthorizer(config.Spec)
+	resourceClusterClient, _ := newClustersClient(azureAuthorizer, config.Spec)
+	clusterState, err := resourceClusterClient.Get(ctx, config.Spec.ResourceGroup, config.Spec.ClusterName)
+
+	// set kubernetes version
+	upstreamVersion := clusterState.KubernetesVersion
+	if upstreamVersion == nil {
+
+		return nil, fmt.Errorf("cannot detect cluster [%s] upstream kubernetes version", config.ClusterName)
+	}
+	upstreamSpec.KubernetesVersion = to.String(upstreamVersion)
+
+	// set tags
+	upstreamSpec.Tags = make(map[string]string)
+	if len(clusterState.Tags) != 0 {
+		upstreamSpec.Tags = to.StringMap(clusterState.Tags)
+	}
+
+	// set agentpool profile
+	for _, ng := range *clusterState.AgentPoolProfiles {
+		var upstreamNG aksv1.NodeGroup
+		upstreamNG.Name = to.String(ng.Name)
+		upstreamNG.Count = to.Int32(ng.Count)
+		upstreamNG.MaxPods = to.Int32(ng.MaxPods)
+		upstreamNG.VMSize = string(ng.VMSize)
+		upstreamNG.OsDiskSizeGB = to.Int32(ng.OsDiskSizeGB)
+		upstreamNG.OsDiskType = string(ng.OsDiskType)
+		upstreamNG.Mode = string(ng.Mode)
+		upstreamNG.OsType = string(ng.OsType)
+		upstreamNG.Version = to.String(ng.OrchestratorVersion)
+		upstreamNG.AvailableZones = to.StringSlice(ng.AvailabilityZones)
+		upstreamSpec.NodeGroups = append(upstreamSpec.NodeGroups, upstreamNG)
+	}
+
+	// set network configuration
+	networkProfile := clusterState.NetworkProfile
+	if networkProfile != nil {
+		upstreamSpec.NetworkPlugin = string(networkProfile.NetworkPlugin)
+		upstreamSpec.NetworkDNSServiceIP = to.String(networkProfile.DNSServiceIP)
+		upstreamSpec.NetworkDockerBridgeCIDR = to.String(networkProfile.DockerBridgeCidr)
+		upstreamSpec.NetworkServiceCIDR = to.String(networkProfile.ServiceCidr)
+		upstreamSpec.NetworkPolicy = string(networkProfile.NetworkPolicy)
+		upstreamSpec.NetworkPodCIDR = to.String(networkProfile.PodCidr)
+		upstreamSpec.LoadBalancerSku = string(networkProfile.LoadBalancerSku)
+	}
+
+	// set linux account profile
+	linuxProfile := clusterState.LinuxProfile
+	if linuxProfile != nil {
+		upstreamSpec.LinuxAdminUsername = to.String(linuxProfile.AdminUsername)
+		sshKeys := *linuxProfile.SSH.PublicKeys
+		upstreamSpec.LinuxSSHPublicKeyContents = to.String(sshKeys[0].KeyData)
+	}
+	return upstreamSpec, err
+}
+
+func updateUpstreamClusterState(ctx context.Context, config *aksv1.AKSClusterConfig, upstreamSpec *aksv1.AKSClusterConfigSpec) (*aksv1.AKSClusterConfig, error) {
```

Comment:
Each fields, once they have been updated to be pointer (where applicable) should be checked for nil value. If they are nil you can skip evaluating them for updates.